src/main/java/com/datadoghq/sketch/ddsketch/store/PaginatedStore.java

```diff
@@ -0,0 +1,271 @@
+package com.datadoghq.sketch.ddsketch.store;
+
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+public final class PaginatedStore implements Store {
+
+    private static final int GROWTH = 8;
+    private static final int PAGE_SIZE = 128;
+    private static final int PAGE_MASK = PAGE_SIZE - 1;
+    private static final int PAGE_SHIFT = Integer.bitCount(PAGE_MASK);
+
+    private double[][] pages = null;
+    private int pageOffset;
+
+    public PaginatedStore() {
+        this(0);
+    }
+
+    PaginatedStore(int pageOffset) {
+        this.pageOffset = pageOffset;
+    }
+
+    PaginatedStore(PaginatedStore store) {
+        this(store.pageOffset);
+        this.pages = deepCopy(store.pages);
+    }
+
+    @Override
+    public boolean isEmpty() {
+        // won't initialise any pages unless until a value is added,
+        // and values can't be removed.
+        return null == pages;
+    }
+
+    @Override
+    public int getMinIndex() {
+        if (null != pages) {
+            for (int i = 0; i < pages.length; ++i) {
+                if (null != pages[i]) {
+                    for (int j = 0; j < pages[i].length; ++j) {
+                        if (pages[i][j] != 0D) {
+                            return ((i - pageOffset) << PAGE_SHIFT) + j;
+                        }
+                    }
+                }
+            }
+        }
+        throw new NoSuchElementException();
+    }
+
+    @Override
+    public int getMaxIndex() {
+        if (null != pages) {
+            for (int i = pages.length - 1; i >= 0; --i) {
+                if (null != pages[i]) {
+                    for (int j = pages[i].length - 1; j >= 0; --j) {
+                        if (pages[i][j] != 0D) {
+                            return ((i - pageOffset) << PAGE_SHIFT) + j;
+                        }
+                    }
+                }
+            }
+        }
+        throw new NoSuchElementException();
+    }
+
+    @Override
+    public void add(int index, double count) {
+        if (count > 0) {
+            int alignedIndex = alignedIndex(index);
+            double[] page = getPage(alignedIndex >>> PAGE_SHIFT);
+            page[alignedIndex & PAGE_MASK] += count;
+        }
+    }
+
+    private double[] getPage(int pageIndex) {
+        double[] page = pages[pageIndex];
+        if (null == page) {
+            page = pages[pageIndex] = new double[PAGE_SIZE];
+        }
+        return page;
+    }
+
+    private int alignedIndex(int index) {
+        int pageIndex = index < 0
+                ? -(-index >>> PAGE_SHIFT) - 1
+                : index >>> PAGE_SHIFT;
```

Comment:
So my brain just starts hurting trying to understand what the `pageIndex` and `pageOffset` represents in this code. I would love some comment describing it broadly.

Reply:
This is much better now with the flipped sign and name change.